### PR TITLE
Create cache for dependent entries when requested first time

### PR DESCRIPTION
**Pull request by ajcvickers**
This greatly improves perf for querying dependents first, followed by principals.

### CODE REVIEW

File: src/EntityFramework.Core/ChangeTracking/Internal/DependentsMap.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            _entityType = foreignKey.DeclaringEntityType;
+            _principalKeyValueFactory = principalKeyValueFactory;
+            _dependentKeyValueFactory = dependentKeyValueFactory;
+            _map = new Dictionary<TKey, List<InternalEntityEntry>>(principalKeyValueFactory.EqualityComparer);
+        }
+
+        public virtual void Add(InternalEntityEntry entry)
+        {
+            TKey key;
+            if (_entityType.IsAssignableFrom(entry.EntityType)
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/ChangeTracking/Internal/DependentsMap.cs]: Should we throw or Assert on `else` ?

File: src/EntityFramework.Core/Metadata/Internal/ForeignKey.cs [hunk 2]
```diff
@@ -465,5 +467,8 @@ private static bool ArePropertyTypesCompatible(IReadOnlyList<IProperty> principa
 
         // Note: This is set and used only by IdentityMapFactoryFactory, which ensures thread-safety
         public virtual object DependentKeyValueFactory { get; set; }
+
+        // Note: This is set and used only by IdentityMapFactoryFactory, which ensures thread-safety
+        public virtual Func<IDependentsMap> DependentsMapFactory { get; set; }
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Internal/ForeignKey.cs]: Implement the interface explicitly?

File: src/EntityFramework.Core/ChangeTracking/Internal/DependentsMap.cs (hunk 1 above)
Comment (ajcvickers) [inline src/EntityFramework.Core/ChangeTracking/Internal/DependentsMap.cs]: Good point. Let me look to see if the check can be made a debug assert.

File: src/EntityFramework.Core/Metadata/Internal/ForeignKey.cs (hunk 2 above)
Comment (ajcvickers) [inline src/EntityFramework.Core/Metadata/Internal/ForeignKey.cs]: Why? It's on internal code already.

Comment (AndriySvyryd) [inline src/EntityFramework.Core/Metadata/Internal/ForeignKey.cs]: To make it even less likely to be used. This is optional, as it doesn't matter much.

**Comment by AndriySvyryd** [comment]
:shipit:

**Comment by ajcvickers** [comment]
Merged

**Pull request closed without merging**